README.md

```diff
@@ -0,0 +1,396 @@
+# What's This?
+
+This repo is an experiment in what a component-based Agent might look like.
+It's not intended to actually behave like an Agent -- all non-component implementation is fake.
+But it is intended to encompass the current design of the Agent, focusing on the different sorts of interconnections between components required.
+
+This provides a loosely-sketched vision of what we will want to get to with the Agent, encouraging discussion now, during the design phase, but with concrete code to look at.
+
+It is also a chance to try out some "interesting" ideas (such as config reducers) that would be much more difficult to sketch out in a real Agent.
+Maybe some of these are bad ideas, maybe they are things we can do later, or maybe we want to do them first thing.
+
+You can see the set of components in [`COMPONENTS.md`](./COMPONENTS.md).
+
+# Building
+
+To build, run `inv build`.
+This will build all binaries in the root of the repository.
+
+To regenerate files in the repo, run `inv generate`.
+
+# Component Guidelines
+
+What follows are draft guidelines for writing components.
+This repository follows these guidelines.
+
+If these guidelines are adopted, then the expectation is that all components would follow them.
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Prerequisites
+
+You should be familiar with [`fx`](https://pkg.go.dev/go.uber.org/fx), the tool used for dependency injection.
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
```

Comment:
I'm thinking foward to the transition to this model -- it eliminates ambiguity as to whether a Go package is a component or not.  Putting all of this in `pkg/` would result in a lot of overlap between component-shaped and non-component-shaped code.  Practically speaking, I think it would be hard to gauge completion of this project if the code overlaps, and I think a lot of "well, it's sort of a component but sort of a global" would sneak in while we weren't looking.  Segregating the component-shaped code means it's easy to find, even for custom linters.

If we wanted to put them all under pkg/components, that'd be fine -- just more to type (and I'm still not sure what "pkg"  means other than "we didn't want to dump all this stuff in the top-level directory").

Reply:
Make senses